### PR TITLE
Added Type hints.

### PR DESCRIPTION
**Pull request by ranbir7**
Type hints added to fit methods.

### CODE REVIEW

**Comment by ranbir7** [comment]
perfect I'll be adding more type hints as well
slowly slowly Im learning more about this project !
Well whats the real purpose of this project? @ConnorStone628

**Comment by ConnorStoneAstro** [comment]
The purpose of this project is to make a system to analyze astronomical images of stars and galaxies and extract information out of them to do science. Stuff like the total brightness, size, colour, orientation, location of the object can be determined by fitting models. There are other programs out there that do this, but none of them work on GPU, or have automatic differentiation, or are written in 100% python. So these features make it unique, fast, and easy to use for the astronomy community.

**Comment by ranbir7** [comment]
> The purpose of this project is to make a system to analyze astronomical images of stars and galaxies and extract information out of them to do science. Stuff like the total brightness, size, colour, orientation, location of the object can be determined by fitting models. There are other programs out there that do this, but none of them work on GPU, or have automatic differentiation, or are written in 100% python. So these features make it unique, fast, and easy to use for the astronomy community.

wow This is Amazing!

**Comment by ConnorStoneAstro** [comment]
Glad you think so! I really think this can be a major tool for astronomers and hopefully something that new surveys use to process millions of galaxies into usable data for science.

File: autoprof/fit/hmc.py [hunk 1]
```diff
@@ -83,15 +83,15 @@ def fit(
...
+    def append_chain(self, state: torch.Tensor)-> None:
         """
         Add a state vector to the MCMC chain
         """
         self.model.set_parameters(state, as_representation = True)
         chain_state = self.model.get_parameter_vector(as_representation = False)
         self.chain.append(chain_state.detach().cpu().clone().numpy())
         
-    def score_fn(self, state: torch.Tensor):
+    def score_fn(self, state: torch.Tensor)-> (float, float):
```
Comment (ConnorStoneAstro) [inline autoprof/fit/hmc.py]: The return values are tensors, so it would be `(torch.Tensor, torch.Tensor)`

**Comment by ranbir7** [comment]
> Glad you think so! I really think this can be a major tool for astronomers and hopefully something that new surveys use to process millions of galaxies into usable data for science.

so what was ur motivation behind making this?

**Comment by ConnorStoneAstro** [comment]
For myself, I have projects related to strong gravitational lensing that will need the automatic differentiation capabilities to do Bayesian statistics on the model parameters in a reasonable amount of time. Currently it can take days to process a single image, but my hope is to bring that down to hours or minutes and get better results. AutoProf is one piece of that puzzle, but there will be more. The project will probably take a couple years to completely finish. In the meantime, I think AutoProf could (hopefully) be used on many other projects as well.

File: autoprof/fit/lm.py [hunk 2]
```diff
@@ -370,38 +371,38 @@ def update_J_AD(self):
...
+    def update_grad(self, Yph)-> None:
         self.grad = torch.matmul(self.J.T, self.W * (self.Y - Yph))
             
     @torch.no_grad()
-    def rho(self, Xp, Xph, h):
+    def rho(self, Xp, Xph, h)-> torch.Tensor:
         return self.ndf*(Xp - Xph) / abs(torch.dot(h, self.L * (torch.abs(torch.diag(self.hess) - self.epsilon5) * h) + self.grad))
 
-    def accept_history(self):
+    def accept_history(self)-> (torch.Tensor, torch.Tensor, List[float]):
```
Comment (ConnorStoneAstro) [inline autoprof/fit/lm.py]: All three will be lists, so the return type is `(List[np.ndarray], List[np.ndarray], List[float])`

File: autoprof/fit/lm.py [hunk 3]
```diff
@@ -412,7 +413,7 @@ def accept_history(self):
                 Ls.append(self.L_history[l])
                 losses.append(self.loss_history[l])
         return lambdas, Ls, losses
-    def progress_history(self):
+    def progress_history(self)-> (torch.Tensor, torch.Tensor, List[float]):
```
Comment (ConnorStoneAstro) [inline autoprof/fit/lm.py]: Same as for accept history `(List[np.ndarray], List[np.ndarray], List[float])`

**Review by ConnorStoneAstro — COMMENTED** [review]
Added some minor comments.

**Comment by ranbir7** [comment]
_Black_ literally is a life saver !
Thanks for mentioning it !

**Comment by ConnorStoneAstro** [comment]
Glad you like it! Black is pretty awesome

File: autoprof/fit/base.py [hunk 4]
```diff
@@ -20,87 +20,106 @@ class BaseOptimizer(object):
...
+        initial_state: Sequence = None,
+        relative_tolerance: float = 1e-3,
+        **kwargs,
+    ) -> None:
+        """
+        Initializes a new instance of the class.
+
+        Args:
+            model (object): An object representing the model.
+            initial_state (Union[None, Tensor]): The initial state of the model could be any tensor.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/base.py]: The type for `initial_state` can be any sequence so it would be `Optional[Sequence]`

File: autoprof/fit/base.py [hunk 5]
```diff
@@ -20,87 +20,106 @@ class BaseOptimizer(object):
...
+
     def res(self) -> np.ndarray:
-        """ Returns the value of lambda (regularization strength) at which minimum chi^2 loss was achieved.
-    
-            Returns:
-            ndarray: Value of lambda at which minimum chi^2 loss was achieved.
+        """Returns the value of lambda (regularization strength) at which minimum chi^2 loss was achieved.
+
+        Returns:
+        ndarray: Value of lambda at which minimum chi^2 loss was achieved.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/base.py]: this line should be indented relative to `Returns:`

File: autoprof/fit/lm.py [hunk 6]
```diff
@@ -351,68 +463,113 @@ def take_low_rho_step(self):
...
     @torch.no_grad()
-    def covariance_matrix(self):
+    def covariance_matrix(self) -> torch.Tensor:
         try:
             return torch.linalg.inv(self.hess)
         except:
-            AP_config.ap_logger.warning("WARNING: Hessian is singular, likely at least one model is non-physical. Will massage Hessian to continue but results should be inspected.")
-            self.hess += torch.eye(len(self.grad), dtype = AP_config.ap_dtype, device = AP_config.ap_device)*(torch.diag(self.hess) == 0)
+            AP_config.ap_logger.warning(
+                "WARNING: Hessian is singular, likely at least one model is non-physical. Will message Hessian to continue but results should be inspected."
```
Comment (ConnorStoneAstro) [inline autoprof/fit/lm.py]: `message` should be `massage`

**Review by ConnorStoneAstro — COMMENTED** [review]
Just a couple minor changes. The black formatting looks great!

File: autoprof/fit/lm.py (hunk 6 above)
Comment (ranbir7) [reply autoprof/fit/lm.py]: I thought it was a typo
cuz massage is a different meaning right?

Comment (ConnorStoneAstro) [reply autoprof/fit/lm.py]: Yes, massage is a different meaning. In this case I am using it to say that I modify the Hessian in a way that makes it not crash the code

**Comment by ranbir7** [comment]
its showing there's some conflict
image
when I click on resolve
these come up
image
image
image

What Shall I do??

**Comment by ConnorStoneAstro** [comment]
Ah, looks like this is because of an update I recently added to main. Looks like we made changes to the same file and git can't figure you how to combine them. All you have to do is go through the highlighted areas and rewrite it to what it should be. For example in the first one you show, it looks like we both added a `from typing import ...` line where one of them gets a `List` object and the other gets a bunch of types. You would want to combine it yourself to be `from typing import List, Callable, Optional, Union, Sequence, Any` so that everything gets included.

**Comment by ranbir7** [comment]
little confused here
image

**Comment by ConnorStoneAstro** [comment]
Looks like there is a system for me to resolve the conflicts. I'll give it a try and let you know how it goes!

**Comment by ranbir7** [comment]
yes please there's alot to resolve !
Im just getting lost.....

**Comment by ConnorStoneAstro** [comment]
Ok, I think I got it, we'll see if the tests run!

**Comment by ranbir7** [comment]
what happened though?

**Comment by ConnorStoneAstro** [comment]
This happens whenever two people edit the same file, git can't figure out how to decide which one is right. As I went through and merged our changes it ended up being 50/50 as to who's code I kept, so git couldn't have made that call on it's own. This is a normal part of code development, though in general I try to avoid it. In this case I had to make changes to the lm.py file for an update I was working on so it couldn't be avoided

**Comment by ranbir7** [comment]
Ooo okayyy
feel free to edit whatever u like !

**Review by ConnorStoneAstro — APPROVED** [review]
looks good to go. I really like the black formatting!

**Comment by ranbir7** [comment]
> looks good to go. I really like the black formatting!

same man !
It's a life saver !

**Pull request merged**